typescript/ts-lowerings/src/org/jetbrains/dukat/tsLowerings/replaceExpressionExtension.kt

```diff
@@ -0,0 +1,140 @@
+    }
+
+    private fun ParameterValueDeclaration.getTypeParameters(): List<ParameterValueDeclaration> {
+        return (this as? TypeDeclaration)?.params ?: emptyList()
+    }
+
+    private fun TopLevelDeclaration?.findAllConstructSignatureDeclarations(): List<ConstructSignatureDeclaration> {
+        return (this as? InterfaceDeclaration)?.let { members.filterIsInstance<ConstructSignatureDeclaration>() }
+            ?: emptyList()
+    }
+
+    private fun TopLevelDeclaration?.getAppliedTypeParametersMap(typeParameters: List<ParameterValueDeclaration>): TypeMapping {
+        val interfaceDeclaration = this as? InterfaceDeclaration ?: return emptyMap()
+        return interfaceDeclaration.typeParameters.withIndex().associate { (index, type) ->
+            val resolvedType = typeParameters.getOrNull(index)
+                ?: type.defaultValue
+                // TODO: Handle multiple constraints
+                ?: type.constraints.firstOrNull()
+
+            type.name to resolvedType!!
+        }
+    }
+
+    private fun List<ConstructSignatureDeclaration>.applyTypeParameters(typeMapping: TypeMapping): List<ConstructSignatureDeclaration> {
+        val typeParametersResolver = ApplyTypeParameters(typeMapping)
+        return map { typeParametersResolver.lowerConstructSignatureDeclaration(it, null) }
+    }
+
+    private fun createClassWith(
+        name: String,
+        uid: String,
+        owner: NodeOwner<ModuleDeclaration>?,
+        constructSignatures: List<ConstructSignatureDeclaration>,
+        staticallyInheritedFrom: TypeDeclaration?
+    ): ClassDeclaration {
+        // TODO: Handle constructor overloading
+        val genericConstructor = constructSignatures.find { it.typeParameters.isNotEmpty() }
+        val constructorDeclarations = constructSignatures.map { it.convertToConstructorDeclaration() }
+        val parentEntities = constructSignatures.map { it.generateHeritageClauseDeclaration() }
+        val staticallyInherited = staticallyInheritedFrom?.generateHeritageClauseDeclaration()
+
+        return ClassDeclaration(
+            uid = uid,
+            name = IdentifierEntity(name),
+            members = constructorDeclarations,
+            typeParameters = genericConstructor?.typeParameters ?: emptyList(),
+            parentEntities = parentEntities,
+            modifiers = setOf(ModifierDeclaration.DECLARE_KEYWORD),
+            staticallyInherited = staticallyInherited?.let { listOf(it) } ?: emptyList(),
+            definitionsInfo = listOf(DefinitionInfoDeclaration(uid, owner?.node?.sourceName ?: "")),
+        )
+    }
+
+    private fun ConstructSignatureDeclaration.generateHeritageClauseDeclaration(): HeritageClauseDeclaration {
+        return (type as? TypeDeclaration).generateHeritageClauseDeclaration()
+    }
+
+    private fun TypeDeclaration?.generateHeritageClauseDeclaration(): HeritageClauseDeclaration {
+       return HeritageClauseDeclaration(
+           extending = false,
```

Comment:
Just to note (not for this review, but overall): now it is used only with `false` value. Maybe it is redundant